<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Integrating genomic mate selection into the NextGen Cassava pipeline</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">implementGMSinCassava</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Preliminary Steps
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01-cleanTPdata.html">Data Cleaning</a>
    </li>
    <li>
      <a href="02-GetBLUPs.html">Get BLUPs</a>
    </li>
    <li>
      <a href="03-validatePedigree.html">Verify pedigree records</a>
    </li>
    <li>
      <a href="04-PreprocessDataFiles.html">Preprocessing input data</a>
    </li>
    <li>
      <a href="08-PHGfiles.html">Preprocess PHG files</a>
    </li>
  </ul>
</li>
<li>
  <a href="05-CrossValidation.html">Cross-validation</a>
</li>
<li>
  <a href="06-GenomicPredictions.html">Genomic (mate) predictions</a>
</li>
<li>
  <a href="07-Results.html">Results</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="about.html">About</a>
    </li>
    <li>
      <a href="license.html">License</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/wolfemd/implementGMSinCassava">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Integrating genomic mate selection into the NextGen Cassava pipeline</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-08-04
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>implementGMSinCassava/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwolfemdimplementGMSinCassavatreef6f58f0cf6da5d3fe555772020a94c3534481d88targetblankf6f58f0a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/wolfemd/implementGMSinCassava/tree/f6f58f0cf6da5d3fe555772020a94c3534481d88" target="_blank">f6f58f0</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwolfemdimplementGMSinCassavatreef6f58f0cf6da5d3fe555772020a94c3534481d88targetblankf6f58f0a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/wolfemd/implementGMSinCassava/tree/f6f58f0cf6da5d3fe555772020a94c3534481d88" target="_blank">f6f58f0</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    code/.DS_Store
    Ignored:    data/.DS_Store

Untracked files:
    Untracked:  analysis/speedUpPredCrossVar.Rmd
    Untracked:  data/AllChrom_RefPanelAndGSprogeny_ReadyForGP_72719.bed
    Untracked:  data/AllChrom_RefPanelAndGSprogeny_ReadyForGP_72719.bim
    Untracked:  data/AllChrom_RefPanelAndGSprogeny_ReadyForGP_72719.fam
    Untracked:  data/AllChrom_RefPanelAndGSprogeny_ReadyForGP_72719.hap.gz
    Untracked:  data/AllChrom_RefPanelAndGSprogeny_ReadyForGP_72719.log
    Untracked:  data/AllChrom_RefPanelAndGSprogeny_ReadyForGP_72719.nosex
    Untracked:  data/AllChrom_RefPanelAndGSprogeny_ReadyForGP_72719.sample
    Untracked:  data/AllChrom_RefPanelAndGSprogeny_ReadyForGP_72719.vcf.gz
    Untracked:  data/DatabaseDownload_2021May04/
    Untracked:  data/blups_forCrossVal.rds
    Untracked:  data/config.txt
    Untracked:  data/config_mw.txt
    Untracked:  data/dosages_IITA_2021May13.rds
    Untracked:  data/dosages_IITA_filtered_2021May13.rds
    Untracked:  data/genmap_2021Aug02.rds
    Untracked:  data/genmap_2021May13.rds
    Untracked:  data/haps_IITA_2021May13.rds
    Untracked:  data/haps_IITA_filtered_2021May13.rds
    Untracked:  data/recombFreqMat_1minus2c_2021May13.rds
    Untracked:  output/Cassava_HMII_V3_Marning_imputation_6-18-21.samples
    Untracked:  output/IITA_CleanedTrialData_2021May10.rds
    Untracked:  output/IITA_ExptDesignsDetected_2021May10.rds
    Untracked:  output/IITA_blupsForModelTraining_twostage_asreml_2021May10.rds
    Untracked:  output/IITA_trials_NOT_identifiable.csv
    Untracked:  output/alphaAssignOutput_iita_pedigree.txt.dams
    Untracked:  output/alphaAssignOutput_iita_pedigree.txt.dams.full
    Untracked:  output/alphaAssignOutput_iita_pedigree.txt.pedigree
    Untracked:  output/alphaAssignOutput_iita_pedigree.txt.pedigree.top
    Untracked:  output/alphaAssignOutput_iita_pedigree.txt.sires
    Untracked:  output/alphaAssignOutput_iita_pedigree.txt.sires.full
    Untracked:  output/crossValPredsA.rds
    Untracked:  output/crossValPredsAD.rds
    Untracked:  output/cvAD_1rep_markerEffects.rds
    Untracked:  output/cvAD_1rep_meanPredAccuracy.rds
    Untracked:  output/cvAD_1rep_parentfolds.rds
    Untracked:  output/cvAD_1rep_predAccuracy.rds
    Untracked:  output/cvAD_1rep_predMeans.rds
    Untracked:  output/cvAD_1rep_predVars.rds
    Untracked:  output/cvAD_1rep_varPredAccuracy.rds
    Untracked:  output/cvAD_5rep5fold_markerEffects.rds
    Untracked:  output/cvAD_5rep5fold_meanPredAccuracy.rds
    Untracked:  output/cvAD_5rep5fold_parentfolds.rds
    Untracked:  output/cvAD_5rep5fold_predMeans.rds
    Untracked:  output/cvAD_5rep5fold_predVars.rds
    Untracked:  output/cvAD_5rep5fold_varPredAccuracy.rds
    Untracked:  output/cvDirDom_1Mb_50kb_pt5_predAccuracy.rds
    Untracked:  output/cvDirDom_1Mb_50kb_pt6_predAccuracy.rds
    Untracked:  output/cvDirDom_1Mb_50kb_pt8_predAccuracy.rds
    Untracked:  output/cvDirDom_1Mb_50kb_pt9_predAccuracy.rds
    Untracked:  output/cvDirDom_5rep5fold_markerEffects.rds
    Untracked:  output/cvDirDom_5rep5fold_meanPredAccuracy.rds
    Untracked:  output/cvDirDom_5rep5fold_parentfolds.rds
    Untracked:  output/cvDirDom_5rep5fold_predMeans.rds
    Untracked:  output/cvDirDom_5rep5fold_predVars.rds
    Untracked:  output/cvDirDom_5rep5fold_varPredAccuracy.rds
    Untracked:  output/cvMeanPredAccuracyA.rds
    Untracked:  output/cvMeanPredAccuracyAD.rds
    Untracked:  output/cvPredMeansA.rds
    Untracked:  output/cvPredMeansAD.rds
    Untracked:  output/cvVarPredAccuracyA.rds
    Untracked:  output/cvVarPredAccuracyAD.rds
    Untracked:  output/genomicMatePredictions_top121parents_ModelAD.csv
    Untracked:  output/genomicMatePredictions_top121parents_ModelAD.rds
    Untracked:  output/genomicMatePredictions_top121parents_ModelDirDom.csv
    Untracked:  output/genomicMatePredictions_top121parents_ModelDirDom.rds
    Untracked:  output/genomicPredictions_ModelAD.csv
    Untracked:  output/genomicPredictions_ModelAD.rds
    Untracked:  output/genomicPredictions_ModelDirDom.csv
    Untracked:  output/genomicPredictions_ModelDirDom.rds
    Untracked:  output/kinship_A_IITA_2021May13.rds
    Untracked:  output/kinship_D_IITA_2021May13.rds
    Untracked:  output/kinship_domGenotypic_IITA_2021July5.rds
    Untracked:  output/markEffsTest.rds
    Untracked:  output/markerEffects.rds
    Untracked:  output/markerEffectsA.rds
    Untracked:  output/markerEffectsAD.rds
    Untracked:  output/maxNOHAV_byStudy.csv
    Untracked:  output/obsCrossMeansAndVars.rds
    Untracked:  output/parentfolds.rds
    Untracked:  output/ped2check_genome.rds
    Untracked:  output/ped2genos.txt
    Untracked:  output/pednames2keep.txt
    Untracked:  output/pednames_Prune100_25_pt25.log
    Untracked:  output/pednames_Prune100_25_pt25.nosex
    Untracked:  output/pednames_Prune100_25_pt25.prune.in
    Untracked:  output/pednames_Prune100_25_pt25.prune.out
    Untracked:  output/potential_dams.txt
    Untracked:  output/potential_sires.txt
    Untracked:  output/predVarTest.rds
    Untracked:  output/samples2keep_IITA_2021May13.txt
    Untracked:  output/samples2keep_IITA_MAFpt01_prune50_25_pt98.log
    Untracked:  output/samples2keep_IITA_MAFpt01_prune50_25_pt98.nosex
    Untracked:  output/samples2keep_IITA_MAFpt01_prune50_25_pt98.prune.in
    Untracked:  output/samples2keep_IITA_MAFpt01_prune50_25_pt98.prune.out
    Untracked:  output/samples2keep_notInPHGdb.txt
    Untracked:  output/stdcvDirDom_1Mb_50kb_pt5_predAccuracy.rds
    Untracked:  output/stdcvDirDom_1Mb_50kb_pt6_predAccuracy.rds
    Untracked:  output/stdcvDirDom_1Mb_50kb_pt8_predAccuracy.rds
    Untracked:  output/stdcvDirDom_1Mb_50kb_pt9_predAccuracy.rds
    Untracked:  output/stdcv_AD_predAccuracy.rds
    Untracked:  output/stdcv_DirDom_predAccuracy.rds
    Untracked:  output/test_cvAD_markerEffects.rds
    Untracked:  output/test_cvAD_meanPredAccuracy.rds
    Untracked:  output/test_cvAD_parentfolds.rds
    Untracked:  output/test_cvAD_predAccuracy.rds
    Untracked:  output/test_cvAD_predMeans.rds
    Untracked:  output/test_cvAD_predVars.rds
    Untracked:  output/test_cvAD_varPredAccuracy.rds
    Untracked:  output/top50crosses.csv
    Untracked:  output/verified_ped.txt

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/index.Rmd</code>) and HTML (<code>docs/index.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/implementGMSinCassava/blob/1a4622b61b6158edadb6fa58c7747799826126dd/analysis/index.Rmd" target="_blank">1a4622b</a>
</td>
<td>
wolfemd
</td>
<td>
2021-08-03
</td>
<td>
Add links to new analyses (marker density vs. accuracy)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/implementGMSinCassava/ba8527c95f93ae9d5aeed916d73079704bfc3c3a/docs/index.html" target="_blank">ba8527c</a>
</td>
<td>
wolfemd
</td>
<td>
2021-08-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/implementGMSinCassava/blob/5e7bc99fde567419c2dcf63bb97b8da878288173/analysis/index.Rmd" target="_blank">5e7bc99</a>
</td>
<td>
wolfemd
</td>
<td>
2021-08-02
</td>
<td>
Results now include both “standard” and parent-wise cross-validation accuracies and plots of cross mean/variance predictions. Ready to knit and publish + push to FTP.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/implementGMSinCassava/769fe811f6ceb34a907efed3f6d9a110ccbc7d03/docs/index.html" target="_blank">769fe81</a>
</td>
<td>
wolfemd
</td>
<td>
2021-07-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/implementGMSinCassava/blob/e176b819b7f370f1b607655daf6357df711d4117/analysis/index.Rmd" target="_blank">e176b81</a>
</td>
<td>
wolfemd
</td>
<td>
2021-07-29
</td>
<td>
Update the project landing page to reflect the near-final state of things. Ready to publish.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/implementGMSinCassava/blob/f8f8a28441247642f057e080daeaea675d8aa8e7/analysis/index.Rmd" target="_blank">f8f8a28</a>
</td>
<td>
wolfemd
</td>
<td>
2021-07-26
</td>
<td>
Update analysis-to-do list. Add placeholders and links about PHG-vs-Beagle comparisons to-come.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/implementGMSinCassava/934141c48eff754aa674925be5357d0a2b8be7d3/docs/index.html" target="_blank">934141c</a>
</td>
<td>
wolfemd
</td>
<td>
2021-07-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/implementGMSinCassava/cc1eb4b5bb7f38ca04846562d27a3bd4951c6c09/docs/index.html" target="_blank">cc1eb4b</a>
</td>
<td>
wolfemd
</td>
<td>
2021-07-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/implementGMSinCassava/blob/772750a8e35e6472113d69e6bc1eff0101aa137f/analysis/index.Rmd" target="_blank">772750a</a>
</td>
<td>
wolfemd
</td>
<td>
2021-07-14
</td>
<td>
DirDom model and selection index calc fully integrated functions.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/implementGMSinCassava/5e45aac791be42c4e432387023b49a41d58ecfe6/docs/index.html" target="_blank">5e45aac</a>
</td>
<td>
wolfemd
</td>
<td>
2021-06-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/implementGMSinCassava/df7a366d1b6d6c3579d2f4fe32935e1cbb7f4de4/docs/index.html" target="_blank">df7a366</a>
</td>
<td>
wolfemd
</td>
<td>
2021-06-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/implementGMSinCassava/blob/c28400faf855cff1720a3566c8acf23641ee2ebf/analysis/index.Rmd" target="_blank">c28400f</a>
</td>
<td>
wolfemd
</td>
<td>
2021-06-10
</td>
<td>
github link added
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/implementGMSinCassava/e66bdadd9561b8bf0e6bd6d77bd02621f05a9341/docs/index.html" target="_blank">e66bdad</a>
</td>
<td>
wolfemd
</td>
<td>
2021-06-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/implementGMSinCassava/blob/a8452bafa0ab7a515967a022ca5c4b2638bb97ce/analysis/index.Rmd" target="_blank">a8452ba</a>
</td>
<td>
wolfemd
</td>
<td>
2021-06-10
</td>
<td>
Initial build of the entire page upon completion of all
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/implementGMSinCassava/blob/8a0c50eca5be95c975c345dc7f41b41863123404/analysis/index.Rmd" target="_blank">8a0c50e</a>
</td>
<td>
wolfemd
</td>
<td>
2021-05-04
</td>
<td>
Start workflowr project.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>We saw in <a href="https://www.biorxiv.org/content/10.1101/2021.01.05.425443v1">an initial study</a>, promising results regarding the prediction of genetic variance in cassava crosses. In that study, we used a high-quality validated pedigree-based phasing pipeline for the marker data. That pipeline is considerably more involved and may not be implementable on the entire breeding germplasm.</p>
<p>In this project, I am integrating cross-variance predictions into our current genomics-enabled breeding pipeline leveraging available data. I conduct additional tests to assess whether current Beagle imputed-and-phased data and a basic pedigree-validation step are sufficient.</p>
<p>This project forms the template for genomic mate selection in NextGen Cassava.</p>
<p><a href="07-Results.html"><strong>See the Results here!</strong></a></p>
<p>We will soon implement in <a href="https://wolfemd.github.io/IITA_2021GS/">summer 2021 for planning fall crossing nurseries and with new GS C5 DArTseqLD data</a>.</p>
<div id="whats-new" class="section level1">
<h1>What’s new?</h1>
<ol style="list-style-type: decimal">
<li><p>Functions previously used in NextGen GS pipeline (<code>code/gsFunctions.R</code>) migrate to and receive an upgrade (<code>code/gmsFunctions.R</code>) to support genomic mate selection. The <code>runGenomicPredictions()</code> function fits genomic mixed-models and produces GEBV/GETGV for existing germplasm as it did previously, but now also returns marker effects and provides a direct input to the new <code>predictCrosses()</code> function, which handles predicting mate selection criteria.</p></li>
<li><p>Support for a directional dominance model (<code>modelType="DirDom"</code>) to incorporate genome-wide homozygosity-effects (inbreeding) into predictions.</p></li>
<li><p>Support for selection indices.</p></li>
<li><p>Improved version of <strong>predCrossVar</strong> functions at <code>code/predCrossVar.R</code></p>
<ul>
<li>eventually to be integrated into the R package</li>
<li>faster unless you want to use the Bayesian approach.</li>
</ul></li>
<li><p>Creation of a single-function interface to accomplish parent-wise cross-validation at <code>code/parentWiseCrossVal.R</code>. Provides estimates of (selection index) accuracy predicting family means and variances.</p></li>
<li><p>Upgraded the “standard” cross-validation function (<code>runCrossVal()</code>), used in previous genomic selection analyses (<em>e.g.</em> <a href="https://wolfemd.github.io/IITA_2020GS/06-Results.html#cross-validation-accuracy">IITA_2020GS CV</a>, <a href="https://wolfemd.github.io/NRCRI_2021GS/05-Results.html#Prediction_accuracy">NRCRI_2021GS CV</a>, <a href="https://wolfemd.github.io/TARI_2020GS/05-Results.html#Prediction_accuracy">TARI_2020GS CV</a>), to measure selection index accuracy (via <code>selInd=</code> and <code>SIwts=</code> arguments) and added a <code>modelType="DirDom"</code> option.</p></li>
</ol>
</div>
<div id="analysis-steps" class="section level1">
<h1>Analysis steps</h1>
<ol style="list-style-type: decimal">
<li><p><a href="01-cleanTPdata.html">Prepare training dataset</a>: Download data from DB, “Clean” and format DB data. Use the standard pipeline to obtain complete breeding trial data for IITA, generate de-regressed BLUPs for downstream analysis.</p>
<ul>
<li><p>Copy <code>gsFunctions.R</code> from <code>code/</code> of most recent NextGen prediction, <a href="https://wolfemd.github.io/NRCRI_2021GS/">NRCRI C3b predicted April 2021</a>.</p></li>
<li><p>Reference <a href="https://wolfemd.github.io/IITA_2020GS/">previous analysis for IITA (2020)</a> in case there are variations.</p></li>
</ul></li>
<li><p><a href="02-GetBLUPs.html">Get BLUPs combining all trial data</a>: Combine data from all trait-trials to get BLUPs for downstream genomic prediction. Fit mixed-model to multi-trial dataset and extract BLUPs, de-regressed BLUPs and weights. Include two rounds of outlier removal.</p></li>
<li><p><a href="03-validatePedigree.html">Validate the pedigree obtained from cassavabase</a>: Before setting up a cross-validation scheme for predictions that depend on a correct pedigree, add a basic verification step to the pipeline. Not trying to fill unknown relationships or otherwise correct the pedigree. Assess evidence that relationship is correct, remove if incorrect.</p></li>
<li><p><a href="04-PreprocessDataFiles.html">Preprocess data files</a>: Prepare haplotype and dosage matrices, GRMs, pedigree and BLUPs, genetic map <em>and</em> recombination frequency matrix, for use in predictions.</p></li>
<li><p><a href="08-PHGfiles.html">Extract and process PHG files</a>: Extract a VCF file from the PHG <code>*.db</code> file produced by Evan Long. Subsequently, prepare haplotype and dosage matrices, GRMs, genetic map <em>and</em> recombination frequency matrix, for use in predictions.</p></li>
<li><p><a href="05-CrossValidation.html">Parent-wise and standard cross-validation</a>:</p>
<ul>
<li><p>All cross-validation analyses are here.</p></li>
<li><p>Features of the new procedures:</p>
<ul>
<li><p>Compute parent-wise cross-validation folds using the validated pedigree. Fit models to get marker effects and make subsequent predictions of cross means and (co)variances.</p></li>
<li><p><strong>Models “AD”</strong> (classic BV+DD partition of additive+dominance) and <strong>“DirDom”</strong> (genotypic add+dom partition with genome-wide homozygosity effect).</p></li>
<li><p>Include use of selection index weights to compute index accuracy.</p></li>
<li><p>Include new models and index predictions in standard fold cross-validation in addition to the parent-wise scheme / function.</p></li>
</ul></li>
<li><p>Additional analyses:</p>
<ul>
<li><a href="https://wolfemd.github.io/implementGMSinCassava/05-CrossValidation.html#Optimize_marker_density_for_speed"><strong>Optimize marker density for speed</strong></a>: Cross-variance prediction is slow, but significant speed gains can be made by using fewer markers for the predictions. Examine the speed benefit vs. accuracy cost trade-off. <a href="https://wolfemd.github.io/implementGMSinCassava/07-Results.html#Optimize_marker_density_for_speed"><strong>Results are here</strong></a>.</li>
</ul></li>
<li><p><input type="checkbox" disabled="" />
<strong><em>[Coming soon]</em></strong> Run cross-validation using both Beagle- vs. PHG-imputed and phased marker data to compare their quality.</p></li>
</ul></li>
<li><p><a href="06-GenomicPredictions.html">Genomic predictions</a>:</p>
<ul>
<li><ol style="list-style-type: upper-alpha">
<li>Standard genomic prediction of individual GEBV and GETGV for all selection candidates using all available data. (B) Predict cross means and variances for genomic mate selection.</li>
</ol></li>
<li><p>Include models “AD” <em>and</em> “DirDom”</p></li>
<li><p>Include prediction of selection index GEBV/GETGV and <span class="math inline">\(UC^{SI}_{parent}\)</span>/<span class="math inline">\(UC^{SI}_{variety}\)</span>.</p></li>
<li><p>New functions at <code>gmsFunctions.R</code> in <code>code/</code></p></li>
</ul></li>
<li><p><a href="07-Results.html">Results</a>: Home for plots and summary tables.</p></li>
</ol>
</div>
<div id="data-and-code-repository-access" class="section level1">
<h1>Data and code repository access</h1>
<p><a href="ftp://ftp.cassavabase.org/marnin_datasets/implementGMSinCassava/">CLICK HERE FOR ACCESS TO THE FULL REPOSITORY</a> (select “Guest” credentials when prompted by the Cassavabase FTP server)</p>
<p>or</p>
<p><a href="https://github.com/wolfemd/implementGMSinCassava/"><strong>DOWNLOAD FROM GitHub*</strong></a></p>
<p>*GitHub only hosts files max 50 Mb.</p>
</div>
<div id="guide-to-key-directories-and-file-names" class="section level1">
<h1>Guide to key directories and file names</h1>
<ol style="list-style-type: decimal">
<li><code>data/</code>: raw data (e.g. unimputed SNP data)</li>
<li><code>output/</code>: outputs (e.g. imputed SNP data)</li>
<li><code>analysis/</code>: most code and workflow documented in <strong>.Rmd</strong> files</li>
<li><code>docs/</code>: compiled <strong>.html</strong>, “knitted” from <strong>.Rmd</strong></li>
<li><code>code/</code>: supporting functions sourced in <code>analysis/*.Rmd</code>’s.</li>
</ol>
<p><strong>FILES OF INTEREST</strong>: everything is in the <code>output/</code> sub-directory (<a href="ftp://ftp.cassavabase.org/marnin_datasets/implementGMSinCassava/output/">click here</a> and select “Guest” credentials when prompted by the Cassavabase FTP server).</p>
<ul>
<li><p><strong>GEBVs for parent selection and GETGVs for variety advancement:</strong></p>
<ul>
<li><p><a href="ftp://ftp.cassavabase.org/marnin_datasets/implementGMSinCassava/output/genomicPredictions_ModelDirDom.csv">download DirDom model genomic predictions</a></p></li>
<li><p><a href="ftp://ftp.cassavabase.org/marnin_datasets/implementGMSinCassava/output/genomicPredictions_ModelAD.csv">download AD model genomic predictions</a></p></li>
</ul></li>
<li><p><strong>Predicted means, variances and usefulness of crosses among top parents:</strong></p>
<ul>
<li><p><a href="ftp://ftp.cassavabase.org/marnin_datasets/implementGMSinCassava/output/genomicMatePredictions_top121parents_ModelDirDom.csv">download DirDom model genomic mate predictions</a></p></li>
<li><p><a href="ftp://ftp.cassavabase.org/marnin_datasets/implementGMSinCassava/output/genomicMatePredictions_top121parents_ModelAD.csv">download AD model genomic mate predictions</a></p></li>
</ul></li>
<li><p>Kinship matrices, dosages, haplotype matrix, recombination frequency matrix, genetic map files</p></li>
</ul>
<p>genomicPredictions_ModelDirDom.csv genomicMatePredictions_top121parents_ModelDirDom.csv genomicPredictions_ModelAD.csv genomicMatePredictions_top121parents_ModelAD.csv</p>
</div>
<div id="analyses-to-do" class="section level1">
<h1>Analyses To Do</h1>
<ol style="list-style-type: decimal">
<li><p><strong>[In Progress]</strong> PHG imputed and phased marker data</p></li>
<li><p>Improve mate selection accuracy by…</p>
<ul>
<li>Choice of genetic map? AWC’s map? Sex specific maps?</li>
<li>Multi-trait REML models, single-step, heterogeneous error and/or of the other usual possibilities,</li>
<li>and/or a return to the compute intense Bayesian posterior distribution approach used in the MS… might be more attractive with fewer markers to make compute faster</li>
</ul></li>
<li><p>Simulation to explore factors impacting estimate of accuracy</p>
<ul>
<li>Impact of phasing switch errors</li>
<li>Quality and quantity of data available on parents</li>
<li>Relatedness of parents</li>
<li>What is the relationship between the true accuracy and estimate of accuracy?</li>
</ul></li>
</ol>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
